hadoop-hdds/docs/content/feature/Snapshot.md

```diff
@@ -23,55 +23,260 @@ summary: Ozone Snapshot
+## **Linux System Configuration**
 
-## Deployment
-----------
-### Cluster and Hardware Configuration
+Proper Linux system configuration and hardware provisioning are important for running Ozone with snapshots in production. The snapshot feature, due to its additional metadata and potential large number of file handles, may require tuning the operating system limits and using high-performance storage for metadata:
 
-Snapshot feature places additional demands on the cluster in terms of CPU, memory and storage. Cluster nodes running Ozone Managers and Ozone Datanodes should be configured with extra storage capacity depending on the number of active snapshots that the user wants to keep. Ozone Snapshots consume an incremental amount of space per snapshot. e.g. if the active object store has 100 GB data (before replication) and a snapshot is taken, then the 100 GB of space will be locked in that snapshot. If the active object store consumes another 10 GB of space (before replication) subsequently then overall space requirement would be 100 GB + 10 GB = 110 GB in total (before replication). This is because common keys between Ozone snapshots and the active object store will share the storage space.
+* **Open File Descriptors Limit:** It is recommended to raise the maximum number of open file descriptors (ulimit) for the Ozone Manager process. Each RocksDB instance (and each snapshot and diff job) can open many files. In production deployments, consider setting the per-process file descriptor limit to **at least 32,000**, and in many cases 64k or 128k is advisable. Ensure the Linux `nofile` ulimit is configured accordingly (via `/etc/security/limits.conf` or systemd service unit limits for ozone).
+* **Use NVMe SSDs for Metadata:** Ozone Manager’s metadata (which includes the RocksDB key tables and snapshot info) should be placed on fast storage. It is highly recommended to use NVMe SSD drives for the OM metadata directories. Snapshots put additional I/O load on OM’s RocksDB (for maintaining multiple snapshot tables and for scanning during diffs). Fast disk (NVMe) improves OM throughput significantly. If NVMe is not available, use the fastest disks possible (SSD over HDD). Additionally, consider RAID1 or RAID10 for OM metadata for reliability, but prioritize random read/write performance.
+* **Memory and CPU for Ozone Manager:** Snapshots consume OM memory proportional to how many are being accessed concurrently. Each active SnapshotDiff job, for example, will load snapshot metadata into memory. By default, OM allows 10 concurrent diff jobs – if each diff is large, this can use considerable memory. Ensure the Ozone Manager is allocated sufficient heap. For a production cluster with snapshots, a larger heap (e.g., 16-32 GB or more) for OM may be warranted, depending on the number of keys and snapshots. Monitor GC and adjust the heap size as needed.
```

Comment:
Diff size being large may not necessarily increase memory usage. Only in the case of FSO when there are a lot of directory objects it may lead to an increase in the memory usage and that also shouldn't be much. We compute snapshot diff by spliing intermediate results to the disk.